lib/internal/modules/esm/default_resolve.js

```diff
@@ -45,12 +45,31 @@ if (experimentalWasmModules)
   extensionFormatMap['.wasm'] = legacyExtensionFormatMap['.wasm'] = 'wasm';
 
 function resolve(specifier, parentURL) {
+  try {
+    const parsed = new URL(specifier);
+    if (parsed.protocol === 'data:') {
+      const [ , mime ] = /^([^/]+\/[^;,]+)(;base64)?,/.exec(parsed.pathname) || [ null, null, null ];
+      const format = ({
+        'text/javascript': 'module',
```

Comment:
text/javascript doesn't necessarily mean Module, it could also mean Script.

Reply:
Under no JS loading spec does Script get checked against the MIME text/javascript. Script is effectively without a MIME and this table matches web standards.

Reply:
I don't think we support Script anywhere, so it would feel weird to do that here. A CommonJS script would be something like `text/vnd.node.js` according to https://github.com/nodejs/TSC/issues/371.

Reply:
CJS is [application/node](https://www.iana.org/assignments/media-types/application/node)

Reply:
then can `application/node` be added to this object?

Reply:
lots of parts of the CJS loader would break when not loaded under a file context.

Reply:
Heh, sorry. Didn't see the new link at the bottom of the issue I linked.

> then can application/node be added to this object?

I'm not convinced there are meaningful semantics for running CJS from a data URL. E.g. there would be no valid value for `__filename` or `__dirname` and no key in the require cache, it would add modules with "invalid" ids into the graph via `module.parent` if it would start requiring other files. Non-`file:` CJS doesn't really fit the model of CJS imo.

Reply:
The `vm` module can't be used to evaluate a non-file CJS module already?

The key in the require cache would be the data URI itself.

Reply:
I think the closest we get is loading `-e` inputs, and that still has to use `process.cwd()` to not break.

Reply:
@ljharb but the require cache works off file paths, not URLs. It also wouldn't make sense for things like `__dirname`, `__filename`, or `module.parent` (if the parent is not a CJS module). I don't think starting to deal with the results of adding a new type of key and potentially parent is a simple task. We would also need to ensure that `require` starts to understand non-file contexts and doesn't load non-builtins, which would be a new thing as well.

Reply:
The expression though — `/^([^/]+\/[^;,]+)(;base64)?,/` seems to assume the presence of mime type before the `[;,]` — regardless of if it is mandatory, it might make sense to test it against long and malformed urls to potentially refine it if necessary.

Sorry for not wanting to muddy this with a bad attempt to wing it here, but I will try to locate the ones I worked on a while back for that very same purpose if it helps.

Reply:
@SMotaal we could, but i can't think of much we could do except limiting the size? Right now this lacks a variety of things, including MIME parameter parsing and the PR for parsing MIMEs is stuck.

Reply:
Since it is anchored, it is certainly possible to add efficient guards in the current expression. I'd like to take on exploring how we can do that here, which is mainly just to carve a limited allowed chars when delimited per spec (I did this a while back just need to dig).

Reply:
@bmeck I looked into the various options for the expression and recommend:

```js
/^(?:((?:text|application)\/(?:[A-Z][-.0-9A-Z]*)?[A-Z]+)((?:;[A-Z][!%'()*\-.0-9A-Z_~]*=[!%'()*\-.0-9A-Z_~]*)*)(;base64)?),/i
```

This would match any `text/` and `application/` subtype, along with the attribute-value parameters like `charset=` (to be parsed separately), and optional `base64` (captured separately from previous parameters).

For now, simply being more restrictive of the character ranges for greedy `*` and `+` captures is likely all we need to avoid unpredictable performance hazards with very long crafted/malformed strings.

See [gist](https://gist.github.com/SMotaal/2e0ede8c6115b828c111fa84a1e952b4) for more details.

Please let me know how to proceed, if this is worth incorporating.

Reply:
you can just post a `suggestion` change and that looks fine